### PR TITLE
fix: create custom Promise.allSettled to work on react native

### PR DESCRIPTION
**Pull request by dpilch**
#### Description of changes

Replace `Promise.allSettled` with a non-fully-spec implementation. In current versions of React Native `Promise.allSettled` is not available resulting in an error. (See https://github.com/aws-amplify/amplify-js/issues/10764#issuecomment-1359993927 for full details).

Another option would be to add a shim from an npm package. However, this would add additional deps.

IMO, the custom implementation serves our needs without introducing additional deps, but I could be convinced to use a shim from npm.

TODO: add a lint rule to prevent the use of `Promise.allSettled`.

#### Issue #, if available

https://github.com/aws-amplify/amplify-js/issues/10764

#### Description of how you validated changes

* Unit test
* Test in react native app

#### Checklist

- [x] PR description included
- [x] `yarn test` passes

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/core/src/Util/BackgroundProcessManager.ts [hunk 1]
```diff
@@ -361,7 +361,7 @@ export class BackgroundProcessManager {
 
 			// Use `allSettled()` because we want to wait for all to finish. We do
 			// not want to stop waiting if there is a failure.
-			this._closingPromise = Promise.allSettled(
+			this._closingPromise = allSettled(
```
Comment (david-mcafee) [inline packages/core/src/Util/BackgroundProcessManager.ts:364]: Thoughts on adding `BackgroundProcessManager.native.ts` alongside this file (as we do for `Reachability`) and only using this implementation for React Native?

File: packages/core/src/Util/BackgroundProcessManager.ts [hunk 2]
```diff
@@ -447,3 +447,30 @@ type JobEntry = {
 	 */
 	description?: string;
 };
+
+/**
+ * Custom non-spec implementation for Promise.allSettled.
```
Comment (david-mcafee) [inline packages/core/src/Util/BackgroundProcessManager.ts:452]: I did a little bit of digging, but I'm sure you've done more than I have - are there spec-compliant implementations we could pull in here?

**Review by david-mcafee — CHANGES_REQUESTED** [review]

File: packages/core/src/Util/BackgroundProcessManager.ts (hunk 2 above)
Comment (dpilch) [reply packages/core/src/Util/BackgroundProcessManager.ts:452]: I know of at least one https://www.npmjs.com/package/promise.allsettled. It just depends on if we want to add an additional dependency or not.

File: packages/core/src/Util/BackgroundProcessManager.ts (hunk 1 above)
Comment (dpilch) [reply packages/core/src/Util/BackgroundProcessManager.ts:364]: I wasn't aware of this option. Is there a way we can easily do this without repeating the other code in this file?

**Pull request closed without merging**